provider/cmd/crd2pulumi/generate.go

```diff
@@ -0,0 +1,209 @@
+)
+
+type CustomResourceGenerator struct {
+	// CustomResourceDefinition represents unmarshalled CRD YAML
+	CustomResourceDefinition unstruct.Unstructured
+	// OutputPath represents the path of the file to generate code in
+	OutputPath string
+	// Language represents the target language to generate code
+	Language string
+	// ApiVersion represents the `apiVersion` field in the CRD YAML
+	APIVersion string
+	// Kind represents the `spec.names.kind` field in the CRD YAML
+	Kind string
+	// Plural represents the `spec.names.plural` field in the CRD YAML
+	Plural string
+	// Group represents the `spec.group` field in the CRD YAML
+	Group string
+	// Versions represents a mapping from each version name in the
+	// `spec.versions` list to its corresponding `openAPIV3Schema` field in the
+	// CRD YAML
+	Versions map[string]map[string]interface{}
+}
+
+func NewCustomResourceGenerator(language, yamlPath, outputPath string) (CustomResourceGenerator, error) {
+	if language != DotNet && language != Go && language != NodeJS && language != Python {
+		return CustomResourceGenerator{}, errors.New("invalid language: " + language)
+	}
+
+	yamlFile, err := ioutil.ReadFile(yamlPath)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("read file %s: %v", yamlPath, err)
+	}
+
+	crd, err := UnmarshalYaml(yamlFile)
+	if err != nil {
+		return CustomResourceGenerator{}, fmt.Errorf("unmarshal %s: %v", yamlPath, err)
+	}
+
+	underscoreFields(crd.Object)
+
+	apiVersion := crd.GetAPIVersion()
+	if apiVersion != v1beta1 && apiVersion != v1 {
+		return CustomResourceGenerator{}, fmt.Errorf("invalid apiVersion %s: %v", apiVersion, err)
+	}
+
+	versions := map[string]map[string]interface{}{}
+
+	validation, foundValidation, _ := unstruct.NestedMap(crd.Object, "spec", "validation", "openAPIV3Schema")
+	if foundValidation { // If present, use the top-level schema to validate all versions
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			versions[name] = validation
+		}
+	} else { // Otherwise use per-version schemas to validate each version
+		versionMaps, _, _ := NestedMapSlice(crd.Object, "spec", "versions")
+		for _, version := range versionMaps {
+			name, _, _ := unstruct.NestedString(version, "name")
+			schema, _, _ := unstruct.NestedMap(version, "schema", "openAPIV3Schema")
+			versions[name] = schema
+		}
+	}
+
+	kind, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "kind")
+	plural, _, _ := unstruct.NestedString(crd.Object, "spec", "names", "plural")
+	group, _, _ := unstruct.NestedString(crd.Object, "spec", "group")
+
+	if outputPath == "" {
+		defaultOutputPath := func(yamlPath, plural, language string) string {
+			var extension string
+			switch language {
+			case NodeJS:
+				extension = "ts"
+			case DotNet:
+				extension = "cs"
+			case Python:
+				extension = "py"
+			case Go:
+				extension = "go"
+			}
```

Comment:
import `"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"`

```suggestion
			default:
				contract.Failf("unexpected language %s", language)
			}
```